[dnm]Of atmos and carn mc: Fastmos has come.

MrsTonedOne opened the pull request:
Do not merge. testing has found a bug where pipenets all delete themselves

I bring fastmos!

Atmos will now fire at a dynamic rate based on how long it's been taking to fire the last few times.
The old fire rate was 2.0 seconds. The new one is a range between 0.5 and 5.0 seconds.

Heres the details:

Subsystems can now have have a dynamic fire rate (defaults to disabled)
This will make the subsystem fire at a rate 3 times slower than that subsystem is currently taking to process. (within given min/max values) eg: if the subsystem is taking 5 deciseconds to process, it's fire rate will become 15 deciseconds.

The air subsystem now uses the new dynamic firerate, min wait is 5ds, max is 50ds.

Moved pipenets over to the air subsystem. Since they ideally should fire in sync with the now dynamic air system.

All atmos machinery (/obj/machinery/atmospherics) is now processed via the air subsystem, not the machinary subsystem for the same reason above

The fire overlay now uses obj pooling to lower lag during fire storms. (fixes #9072)

MrsTonedOne committed:
Minor rewrite of atmos and atmos subsystems

Sub systems can now have have a dynamic fire rate
This will make the sub system fire at a rate 3 times slower than that subsystem is currently taking to process. (within giving min/max values)

The air subsystem uses dynamic firerate, min wait is 5ds, max is 50ds.

Moved pipenets over to the air subsystem. Since they ideally should fire in sync with the now dynamic air system.
All atmos machinery (/obj/machinery/atmospherics) is now processed via the air subsystem, not the machinary subsystem for the same reason above

The fire overlay now uses obj pooling.

Alek2ander commented:
Praise!

MrsTonedOne committed:
fixes mc processing rate typo

RemieRichards commented:
Amazing :+1: 
(Except for "mechinery" it's machinery)

AlexanderUlanH commented:
We didn't have this already?

Supermichael777 commented:
no curently we have atmos: lagpocalysp edition. somthing to do with the singularity we think. or lighting being an object is causing problems with our lord and savior but that would be to easy.

Firecage commented:
I am pretty sure the atmos we had now isn't lagpocalypse, and was way faster than the atmos before that. So this one will be super fast.

MrStonedOne commented:
@Firecage @Supermichael777 @AlexanderUlanH What we have on the servers was never added to the github, it was a 1 week trial on the servers that never got reverted because it turned out to not be that bad.

This is me adding fastmos to the codebase, and giving it something more, I fixed vents and pipes not running fast (you will notice that on the servers) and I fixed fire being a lag fest(some what). and I made atmos automatically slow down when it's lagging up the server.

None of that is on the server, what is on the server is air movement running at 0.5 seconds.

This just formalizes that into something that works better for both us and other code bases.

optimumtact commented:
Nice work on that pooling
:+1:

MrPerson commented on code/LINDA/LINDA_turf_tile.dm:
```diff
@@ -274,11 +274,11 @@ atom/movable/var/pressure_resistance = 5
 
 atom/movable/proc/experience_pressure_difference(pressure_difference, direction)
-	if(last_forced_movement >= SSair.times_fired)
+	if(last_forced_movement >= world.time-10)
```
Is that 10 just some magic number or does it actually mean something?

MrStonedOne replied:
From my understanding, 1 second.

This keeps fastmos from pulling people out of an air lock twice a second.

Is there a better way to do this?

MrPerson replied:
No that's fine

MrPerson commented on code/controllers/master_controller.dm:
```diff
@@ -23,19 +23,20 @@ var/global/datum/controller/game_controller/master_controller = new()
-
+	if(GCD < 1)
+		GCD = 1
+	processing_interval = GCD
```
This whole block is entirely unnecessary. And frankly it shouldn't be 1. The lowest value should be world.tick_lag. The reason for this is simple: if we go to 20 fps or higher, it would be nice to take advantage of every other tick, which would currently be wasted.

MrStonedOne replied:
This was from carnmc mostly, i just made into a proc instead of being inside mc's new.

What this does, is sets the mc's sleep rate to be the lowest it needs to be to hit every task on time. Because of the fact that subsystems can change how often they run, this is dynamically updated only when it matters.

If every subsystem is either 5ds, 10ds or 20ds, than the mc only needs to run every 5ds. but if one of those is 7ds, than the mc needs to run every 1ds to be able to hit that 7ds one on time.

Its set to 1 because a note in carnmc on subsystems.dm says wait should only be positive whole numbers. I'm not quite at the point where I understand carnmc well enough to solve issues with partial numbers, and the changes I've made to carnmc are already borderlining the scope of what should be in one pr.

Anywho, the only thing new I've coded in this pr is hot spot pooling and the logic around changing the wait of a subsystem, everything else was just moving existing code around and generalizing existing code to work with dynamic fastmos.

MrPerson replied:
The only reason it wants whole numbers is so the gcd above works. What I'm saying is that by removing the gcd, you can use any value you want.

Yeah it's out of scope and all that, it's more just a general note.

MrStonedOne commented:
Do not merge. testing has found a bug where pipenets all delete themselves

RemieRichards commented:
what the

Jalleo commented:
PIPENETS!

MrStonedOne commented:
I'm gonna have to do a bit more of a carnmc rewrite to make this work bug free than i thought.

So i figured, I'll just turn this into a carnmc and subsystem minor rewrite.

Firecage commented:
@MrStonedOne Oh, I thought the atmos that was on the server was still LINDA.

optimumtact commented:
it is, this isn't an atmos rewrite, just changes the tick speed to be faster and dynamic
